[CI] InternalAutoDateHistogramTests::testReduceRandom fails reproducable for certain seed

The following test failed in a PR-build and is reproducible for me on my local box:


```
./gradlew :server:unitTest -Dtests.seed=24D81295C2796C6C -Dtests.class=org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests -Dtests.method="testReduceRandom" -Dtests.security.manager=true -Dtests.locale=el-CY -Dtests.timezone=Europe/Nicosia -Dcompiler.java=11 -Druntime.java=8
```

fails with:

```
  1> [2019-02-28T10:25:36,507][INFO ][o.e.s.a.b.h.InternalAutoDateHistogramTests] [testReduceRandom] before test
  1> [2019-02-28T10:25:36,723][INFO ][o.e.s.a.b.h.InternalAutoDateHistogramTests] [testReduceRandom] after test
  2> REPRODUCE WITH: ./gradlew :server:unitTest -Dtests.seed=24D81295C2796C6C -Dtests.class=org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests -Dtests.method="testReduceRandom" -Dtests.security.manager=true -Dtests.locale=el-CY -Dtests.timezone=Europe/Nicosia -Dcompiler.java=11 -Druntime.java=8
FAILURE 1.13s | InternalAutoDateHistogramTests.testReduceRandom <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: expected:<{1548979200000=1700}> but was:<{1548979200000=1569, 1551398400000=131}>
   >    at __randomizedtesting.SeedInfo.seed([24D81295C2796C6C:E4E5BE84E424F53A]:0)
   >    at org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests.assertReduced(InternalAutoDateHistogramTests.java:200)
   >    at org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests.assertReduced(InternalAutoDateHistogramTests.java:50)
   >    at org.elasticsearch.test.InternalAggregationTestCase.testReduceRandom(InternalAggregationTestCase.java:294)
   >    at java.lang.Thread.run(Thread.java:748)
  2> NOTE: leaving temporary files on disk at: /home/hendrik/work/git-elastic/elasticsearch/server/build/testrun/unitTest/J0/temp/org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests_24D81295C2796C6C-001
  2> NOTE: test params are: codec=Asserting(Lucene80): {}, docValues:{}, maxPointsInLeafNode=1501, maxMBSortInHeap=7.215678116490036, sim=Asserting(org.apache.lucene.search.similarities.AssertingSimilarity@5f67521f), locale=el-CY, timezone=Europe/Nicosia
  2> NOTE: Linux 4.18.0-15-generic amd64/Oracle Corporation 1.8.0_191 (64-bit)/cpus=4,threads=1,free=449796568,total=514850816
  2> NOTE: All tests run in this JVM: [InternalAutoDateHistogramTests]
Completed [1/1] in 2.83s, 1 test, 1 failure <<< FAILURES!


> Task :server:unitTest FAILED
```

Tested only on master but likely applies to other versions.

Also fails for 

```
./gradlew :server:unitTest -Dtests.seed=7DBCFCFC5EFB7BF8 -Dtests.class=org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests -Dtests.method="testReduceRandom" -Dtests.security.manager=true -Dtests.locale=fil-PH -Dtests.timezone=Asia/Krasnoyarsk -Dcompiler.java=11 -Druntime.java=11
```

```
➜  elasticsearch git:(master) ./gradlew :server:unitTest -Dtests.seed=7DBCFCFC5EFB7BF8 -Dtests.class=org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests -Dtests.method="testReduceRandom" -Dtests.security.manager=true -Dtests.locale=fil-PH -Dtests.timezone=Asia/Krasnoyarsk -Dcompiler.java=11 -Druntime.java=11

> Configure project :benchmarks
=======================================
Elasticsearch Build Hamster says Hello!
  Gradle Version        : 5.2.1
  OS Info               : Linux 4.18.0-15-generic (amd64)
  JDK Version           : 11 (Oracle Corporation 11.0.1 [OpenJDK 64-Bit Server VM 11.0.1+13-Ubuntu-3ubuntu3.18.10.1])
  JAVA_HOME             : /usr/lib/jvm/java-11-openjdk-amd64
  Random Testing Seed   : 7DBCFCFC5EFB7BF8
=======================================
Suite: org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests
  1> [2019-03-01T01:57:18,664][INFO ][o.e.s.a.b.h.InternalAutoDateHistogramTests] [testReduceRandom] before test
  1> [2019-03-01T01:57:18,762][INFO ][o.e.s.a.b.h.InternalAutoDateHistogramTests] [testReduceRandom] after test
  2> REPRODUCE WITH: ./gradlew :server:unitTest -Dtests.seed=7DBCFCFC5EFB7BF8 -Dtests.class=org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests -Dtests.method="testReduceRandom" -Dtests.security.manager=true -Dtests.locale=fil-PH -Dtests.timezone=Asia/Krasnoyarsk -Dcompiler.java=11 -Druntime.java=11
FAILURE 0.41s | InternalAutoDateHistogramTests.testReduceRandom <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: expected:<{1548979200000=989}> but was:<{1548979200000=928, 1551398400000=61}>
   >    at __randomizedtesting.SeedInfo.seed([7DBCFCFC5EFB7BF8:BD8150ED78A6E2AE]:0)
   >    at org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests.assertReduced(InternalAutoDateHistogramTests.java:200)
   >    at org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests.assertReduced(InternalAutoDateHistogramTests.java:50)
   >    at org.elasticsearch.test.InternalAggregationTestCase.testReduceRandom(InternalAggregationTestCase.java:294)
   >    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   >    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
   >    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   >    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
   >    at java.base/java.lang.Thread.run(Thread.java:834)
  2> NOTE: leaving temporary files on disk at: /home/brownbear/src/elasticsearch/server/build/testrun/unitTest/J0/temp/org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests_7DBCFCFC5EFB7BF8-001
  2> NOTE: test params are: codec=Asserting(Lucene80): {}, docValues:{}, maxPointsInLeafNode=1350, maxMBSortInHeap=7.054482784277555, sim=Asserting(org.apache.lucene.search.similarities.AssertingSimilarity@73807316), locale=fil-PH, timezone=Asia/Krasnoyarsk
  2> NOTE: Linux 4.18.0-15-generic amd64/Oracle Corporation 11.0.1 (64-bit)/cpus=16,threads=1,free=502433536,total=536870912
  2> NOTE: All tests run in this JVM: [InternalAutoDateHistogramTests]
Completed [1/1] in 1.18s, 1 test, 1 failure <<< FAILURES!


> Task :server:unitTest FAILED
   [junit4] <JUnit4> says g'day! Master seed: 7DBCFCFC5EFB7BF8
==> Test Info: seed=7DBCFCFC5EFB7BF8; jvm=1; suite=1
Tests with failures:
  - org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests.testReduceRandom

   [junit4] JVM J0:     0.26 ..     1.76 =     1.51s
   [junit4] Execution time total: 1.77 sec.
   [junit4] Tests summary: 1 suite, 1 test, 1 failure

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':server:unitTest'.
> There were test failures: 1 suite, 1 test, 1 failure [seed: 7DBCFCFC5EFB7BF8]

```
➜  elasticsearch git:(master) ./gradlew :server:unitTest -Dtests.seed=7DBCFCFC5EFB7BF8 -Dtests.class=org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests -Dtests.method="testReduceRandom" -Dtests.security.manager=true -Dtests.locale=fil-PH -Dtests.timezone=Asia/Krasnoyarsk -Dcompiler.java=11 -Druntime.java=11

> Configure project :benchmarks
=======================================
Elasticsearch Build Hamster says Hello!
  Gradle Version        : 5.2.1
  OS Info               : Linux 4.18.0-15-generic (amd64)
  JDK Version           : 11 (Oracle Corporation 11.0.1 [OpenJDK 64-Bit Server VM 11.0.1+13-Ubuntu-3ubuntu3.18.10.1])
  JAVA_HOME             : /usr/lib/jvm/java-11-openjdk-amd64
  Random Testing Seed   : 7DBCFCFC5EFB7BF8
=======================================
Suite: org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests
  1> [2019-03-01T01:57:18,664][INFO ][o.e.s.a.b.h.InternalAutoDateHistogramTests] [testReduceRandom] before test
  1> [2019-03-01T01:57:18,762][INFO ][o.e.s.a.b.h.InternalAutoDateHistogramTests] [testReduceRandom] after test
  2> REPRODUCE WITH: ./gradlew :server:unitTest -Dtests.seed=7DBCFCFC5EFB7BF8 -Dtests.class=org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests -Dtests.method="testReduceRandom" -Dtests.security.manager=true -Dtests.locale=fil-PH -Dtests.timezone=Asia/Krasnoyarsk -Dcompiler.java=11 -Druntime.java=11
FAILURE 0.41s | InternalAutoDateHistogramTests.testReduceRandom <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: expected:<{1548979200000=989}> but was:<{1548979200000=928, 1551398400000=61}>
   >    at __randomizedtesting.SeedInfo.seed([7DBCFCFC5EFB7BF8:BD8150ED78A6E2AE]:0)
   >    at org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests.assertReduced(InternalAutoDateHistogramTests.java:200)
   >    at org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests.assertReduced(InternalAutoDateHistogramTests.java:50)
   >    at org.elasticsearch.test.InternalAggregationTestCase.testReduceRandom(InternalAggregationTestCase.java:294)
   >    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   >    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
   >    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   >    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
   >    at java.base/java.lang.Thread.run(Thread.java:834)
  2> NOTE: leaving temporary files on disk at: /home/brownbear/src/elasticsearch/server/build/testrun/unitTest/J0/temp/org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests_7DBCFCFC5EFB7BF8-001
  2> NOTE: test params are: codec=Asserting(Lucene80): {}, docValues:{}, maxPointsInLeafNode=1350, maxMBSortInHeap=7.054482784277555, sim=Asserting(org.apache.lucene.search.similarities.AssertingSimilarity@73807316), locale=fil-PH, timezone=Asia/Krasnoyarsk
  2> NOTE: Linux 4.18.0-15-generic amd64/Oracle Corporation 11.0.1 (64-bit)/cpus=16,threads=1,free=502433536,total=536870912
  2> NOTE: All tests run in this JVM: [InternalAutoDateHistogramTests]
Completed [1/1] in 1.18s, 1 test, 1 failure <<< FAILURES!


> Task :server:unitTest FAILED
   [junit4] <JUnit4> says g'day! Master seed: 7DBCFCFC5EFB7BF8
==> Test Info: seed=7DBCFCFC5EFB7BF8; jvm=1; suite=1
Tests with failures:
  - org.elasticsearch.search.aggregations.bucket.histogram.InternalAutoDateHistogramTests.testReduceRandom

   [junit4] JVM J0:     0.26 ..     1.76 =     1.51s
   [junit4] Execution time total: 1.77 sec.
   [junit4] Tests summary: 1 suite, 1 test, 1 failure

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':server:unitTest'.
> There were test failures: 1 suite, 1 test, 1 failure [seed: 7DBCFCFC5EFB7BF8]

```

--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregationBuilder.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregationBuilder.java
@@ -298,5 +298,10 @@ public class AutoDateHistogramAggregationBuilder
                 && Objects.equals(dateTimeUnit, other.dateTimeUnit)
                 ;
         }
+
+        @Override
+        public String toString() {
+            return "RoundingInfo[" + rounding + " " + Arrays.toString(innerIntervals) + "]";
+        }
     }
 }

--- a/server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalAutoDateHistogramTests.java
+++ b/server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalAutoDateHistogramTests.java
@@ -21,6 +21,7 @@ package org.elasticsearch.search.aggregations.bucket.histogram;
 
 import org.elasticsearch.common.Rounding;
 import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.time.DateFormatter;
 import org.elasticsearch.search.DocValueFormat;
 import org.elasticsearch.search.aggregations.InternalAggregations;
 import org.elasticsearch.search.aggregations.ParsedMultiBucketAggregation;
@@ -40,21 +41,23 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.TreeMap;
+import java.util.concurrent.TimeUnit;
 
-import static org.elasticsearch.common.unit.TimeValue.timeValueHours;
-import static org.elasticsearch.common.unit.TimeValue.timeValueMinutes;
-import static org.elasticsearch.common.unit.TimeValue.timeValueSeconds;
 import static org.hamcrest.Matchers.equalTo;
 
 public class InternalAutoDateHistogramTests extends InternalMultiBucketAggregationTestCase<InternalAutoDateHistogram> {
 
     private DocValueFormat format;
     private RoundingInfo[] roundingInfos;
+    private long defaultStart;
+    private int roundingIndex;
 
     @Override
     public void setUp() throws Exception {
         super.setUp();
         format = randomNumericDocValueFormat();
+        defaultStart = randomLongBetween(0, DateFormatter.forPattern("date_optional_time").parseMillis("2050-01-01"));
+        roundingIndex = between(0, AutoDateHistogramAggregationBuilder.buildRoundings(null, null).length - 1);
     }
 
     @Override
@@ -68,10 +71,13 @@ public class InternalAutoDateHistogramTests extends InternalMultiBucketAggregati
         int targetBuckets = randomIntBetween(1, nbBuckets * 2 + 1);
         List<InternalAutoDateHistogram.Bucket> buckets = new ArrayList<>(nbBuckets);
 
-        long startingDate = System.currentTimeMillis();
+        long startingDate = defaultStart;
+        if (rarely()) {
+            startingDate += randomFrom(TimeUnit.MINUTES, TimeUnit.HOURS, TimeUnit.DAYS).toMillis(between(1, 10000));
+        }
 
         long interval = randomIntBetween(1, 3);
-        long intervalMillis = randomFrom(timeValueSeconds(interval), timeValueMinutes(interval), timeValueHours(interval)).getMillis();
+        long intervalMillis = roundingInfos[roundingIndex].roughEstimateDurationMillis * interval;
 
         for (int i = 0; i < nbBuckets; i++) {
             long key = startingDate + (intervalMillis * i);
@@ -108,10 +114,6 @@ public class InternalAutoDateHistogramTests extends InternalMultiBucketAggregati
         assertThat(result, equalTo(2));
     }
 
-    public void testReduceRandom() {
-        super.testReduceRandom();
-    }
-
     @Override
     protected void assertReduced(InternalAutoDateHistogram reduced, List<InternalAutoDateHistogram> inputs) {
 
@@ -134,41 +136,49 @@ public class InternalAutoDateHistogramTests extends InternalMultiBucketAggregati
         RoundingInfo roundingInfo = roundingInfos[roundingIndex];
 
         long normalizedDuration = (highest - lowest) / roundingInfo.getRoughEstimateDurationMillis();
-        long innerIntervalToUse = roundingInfo.innerIntervals[0];
         int innerIntervalIndex = 0;
 
-        // First, try to calculate the correct innerInterval using the normalizedDuration.
-        // This handles cases where highest and lowest are further apart than the interval being used.
+        /*
+         * Guess the interval to use based on the roughly estimated
+         * duration. It'll be accurate or it'll produce more buckets
+         * than we need but it is quick. 
+         */
         if (normalizedDuration != 0) {
             for (int j = roundingInfo.innerIntervals.length-1; j >= 0; j--) {
                 int interval = roundingInfo.innerIntervals[j];
                 if (normalizedDuration / interval < reduced.getBuckets().size()) {
-                    innerIntervalToUse = interval;
                     innerIntervalIndex = j;
                 }
             }
         }
 
-        long intervalInMillis = innerIntervalToUse * roundingInfo.getRoughEstimateDurationMillis();
-        int bucketCount = getBucketCount(lowest, highest, roundingInfo, intervalInMillis);
-
-        //Next, if our bucketCount is still above what we need, we'll go back and determine the interval
-        // based on a size calculation.
-        if (bucketCount > reduced.getBuckets().size()) {
-            for (int i = innerIntervalIndex; i < roundingInfo.innerIntervals.length; i++) {
-                long newIntervalMillis = roundingInfo.innerIntervals[i] * roundingInfo.getRoughEstimateDurationMillis();
-                if (getBucketCount(lowest, highest, roundingInfo, newIntervalMillis) <= reduced.getBuckets().size()) {
-                    innerIntervalToUse = roundingInfo.innerIntervals[i];
-                    intervalInMillis = innerIntervalToUse * roundingInfo.getRoughEstimateDurationMillis();
-                }
+        /*
+         * Next pick smaller intervals until we find the one that makes the right
+         * number of buckets.
+         */
+        int innerIntervalToUse;
+        do {
+            innerIntervalToUse = roundingInfo.innerIntervals[innerIntervalIndex];
+            int bucketCount = getBucketCount(lowest, highest, roundingInfo.rounding, innerIntervalToUse);
+            if (bucketCount == reduced.getBuckets().size()) {
+                break;
             }
-        }
+            if (bucketCount < reduced.getBuckets().size()) {
+                innerIntervalToUse = roundingInfo.innerIntervals[Math.max(0, innerIntervalIndex - 1)];
+                break;
+            }
+        } while (++innerIntervalIndex < roundingInfo.innerIntervals.length);
 
-        Map<Long, Long> expectedCounts = new TreeMap<>();
-        for (long keyForBucket = roundingInfo.rounding.round(lowest);
-             keyForBucket <= roundingInfo.rounding.round(highest);
-             keyForBucket = keyForBucket + intervalInMillis) {
-            expectedCounts.put(keyForBucket, 0L);
+        assertThat(reduced.getInterval().toString(), equalTo(innerIntervalToUse + roundingInfo.unitAbbreviation));
+        Map<Instant, Long> expectedCounts = new TreeMap<>();
+        long keyForBucket = roundingInfo.rounding.round(lowest);
+        while (keyForBucket <= roundingInfo.rounding.round(highest)) {
+            long nextKey = keyForBucket;
+            for (int i = 0; i < innerIntervalToUse; i++) {
+                nextKey = roundingInfo.rounding.nextRoundingValue(nextKey);
+            }
+            Instant key = Instant.ofEpochMilli(keyForBucket);
+            expectedCounts.put(key, 0L);
 
             // Iterate through the input buckets, and for each bucket, determine if it's inside
             // the range of the bucket in the outer loop. if it is, add the doc count to the total
@@ -178,26 +188,26 @@ public class InternalAutoDateHistogramTests extends InternalMultiBucketAggregati
                 for (Histogram.Bucket bucket : histogram.getBuckets()) {
                     long roundedBucketKey = roundingInfo.rounding.round(((ZonedDateTime) bucket.getKey()).toInstant().toEpochMilli());
                     long docCount = bucket.getDocCount();
-                    if (roundedBucketKey >= keyForBucket
-                        && roundedBucketKey < keyForBucket + intervalInMillis) {
-                        expectedCounts.compute(keyForBucket,
-                            (key, oldValue) -> (oldValue == null ? 0 : oldValue) + docCount);
+                    if (roundedBucketKey >= keyForBucket && roundedBucketKey < nextKey) {
+                        expectedCounts.compute(key,
+                            (k, oldValue) -> (oldValue == null ? 0 : oldValue) + docCount);
                     }
                 }
             }
+            keyForBucket = nextKey;
         }
 
         // If there is only a single bucket, and we haven't added it above, add a bucket with no documents.
         // this step is necessary because of the roundedBucketKey < keyForBucket + intervalInMillis above.
         if (roundingInfo.rounding.round(lowest) == roundingInfo.rounding.round(highest) && expectedCounts.isEmpty()) {
-            expectedCounts.put(roundingInfo.rounding.round(lowest), 0L);
+            expectedCounts.put(Instant.ofEpochMilli(roundingInfo.rounding.round(lowest)), 0L);
         }
 
 
         // pick out the actual reduced values to the make the assertion more readable
-        Map<Long, Long> actualCounts = new TreeMap<>();
+        Map<Instant, Long> actualCounts = new TreeMap<>();
         for (Histogram.Bucket bucket : reduced.getBuckets()) {
-            actualCounts.compute(((ZonedDateTime) bucket.getKey()).toInstant().toEpochMilli(),
+            actualCounts.compute(((ZonedDateTime) bucket.getKey()).toInstant(),
                     (key, oldValue) -> (oldValue == null ? 0 : oldValue) + bucket.getDocCount());
         }
         assertEquals(expectedCounts, actualCounts);
@@ -211,11 +221,13 @@ public class InternalAutoDateHistogramTests extends InternalMultiBucketAggregati
         assertThat(reduced.getInterval(), equalTo(expectedInterval));
     }
 
-    private int getBucketCount(long lowest, long highest, RoundingInfo roundingInfo, long intervalInMillis) {
+    private int getBucketCount(long min, long max, Rounding rounding, int interval) {
         int bucketCount = 0;
-        for (long keyForBucket = roundingInfo.rounding.round(lowest);
-             keyForBucket <= roundingInfo.rounding.round(highest);
-             keyForBucket = keyForBucket + intervalInMillis) {
+        long key = rounding.round(min);
+        while (key < max) {
+            for (int i = 0; i < interval; i++) {
+                key = rounding.nextRoundingValue(key);
+            }
             bucketCount++;
         }
         return bucketCount;

